server/src/main/java/org/elasticsearch/action/admin/indices/alias/TransportIndicesAliasesAction.java

```diff
@@ -98,55 +98,59 @@ protected void masterOperation(Task task, final IndicesAliasesRequest request, f
                     case ADD:
+                        // Fail if parameters are used that data stream aliases don't support:
+                        if (action.filter() != null) {
+                            throw new IllegalArgumentException("aliases that point to data streams don't support filters");
+                        }
+                        if (action.routing() != null) {
+                            throw new IllegalArgumentException("aliases that point to data streams don't support routing");
+                        }
+                        if (action.indexRouting() != null) {
+                            throw new IllegalArgumentException("aliases that point to data streams don't support index_routing");
+                        }
+                        if (action.searchRouting() != null) {
+                            throw new IllegalArgumentException("aliases that point to data streams don't support search_routing");
+                        }
+                        if (action.isHidden() != null) {
+                            throw new IllegalArgumentException("aliases that point to data streams don't support is_hidden");
+                        }
+                        // Fail if expressions match both data streams and regular indices:
+                        if (nonBackingIndices.isEmpty() == false) {
+                            throw new IllegalArgumentException("expressions " + Arrays.toString(action.indices()) +
+                                " that match with both data streams and regular indices are disallowed");
+                        }
                         for (String dataStreamName : concreteDataStreams) {
                             for (String alias : concreteDataStreamAliases(action, state.metadata(), dataStreamName)) {
                                 finalActions.add(new AliasAction.AddDataStreamAlias(alias, dataStreamName, action.writeIndex()));
                             }
                         }
-                        break;
+                        continue;
                     case REMOVE:
                         for (String dataStreamName : concreteDataStreams) {
                             for (String alias : concreteDataStreamAliases(action, state.metadata(), dataStreamName)) {
                                 finalActions.add(
                                     new AliasAction.RemoveDataStreamAlias(alias, dataStreamName, action.mustExist()));
                             }
                         }
-                        break;
+                        if (nonBackingIndices.isEmpty() == false) {
+                            // Regular aliases/indices match as well with the provided expression.
+                            // (Only when adding new aliases, matching both data streams and indices is disallowed)
+                            break;
+                        } else {
+                            continue;
+                        }
```

Comment:
I didn't follow the logic here. Do the `break` and `continue` result in different outcomes depending on the value of `nonBackingIndices.isEmpty()`?

Reply:
Yes. The `noneBackingIndices` set contains the names of regular indices that match with this alias action. In that case, this transport action should attempt to remove indices aliases.

If the set is empty then the logic should stop here, because otherwise resolving to concrete indices that happens a bit later would fail. (since no regular indices would be resolved)